src/transformers/models/rwkv5/tokenization_rwkv5.py

```diff
@@ -0,0 +1,545 @@
+        self.ch = ch
+        self.to = [None for ch in range(256)]
+        self.values = set()
+        self.front = front
+
+    def __repr__(self):
+        fr = self
+        ret = []
+        while fr is not None:
+            if fr.ch is not None:
+                ret.append(fr.ch)
+            fr = fr.front
+        return "<TRIE %s %s>" % (ret[::-1], self.values)
+
+    def add(self, key: bytes, idx: int = 0, val=None):
+        if idx == len(key):
+            if val is None:
+                val = key
+            self.values.add(val)
+            return self
+        ch = key[idx]
+        if self.to[ch] is None:
+            self.to[ch] = TRIE(front=self, ch=ch)
+        return self.to[ch].add(key, idx=idx + 1, val=val)
+
+    def find_longest(self, key: bytes, idx: int = 0):
+        u: TRIE = self
+        ch: int = key[idx]
+
+        while u.to[ch] is not None:
+            u = u.to[ch]
+            idx += 1
+            if u.values:
+                ret = idx, u, u.values
+            if idx == len(key):
+                break
+            ch = key[idx]
+        return ret
+
+
+class RWKVWorldTokenizer(PreTrainedTokenizer):
+    vocab_files_names = VOCAB_FILES_NAMES
+    model_input_names = ["input_ids", "attention_mask"]
+
+    def __init__(self, vocab_file, errors="replace", pad_token="0", **kwargs):
+        self.add_bos_token = False
+        self.encoder = {}
+        sorted = []  # must be already sorted
+        with open(vocab_file, "r", encoding="utf-8") as f:
+            lines = f.readlines()
+        for l in lines:
+            idx = int(l[: l.index(" ")])
+            x = eval(l[l.index(" ") : l.rindex(" ")])
+            x = x.encode("utf-8") if isinstance(x, str) else x
+            assert isinstance(x, bytes)
+            assert len(x) == int(l[l.rindex(" ") :])
+            sorted += [x]
+            self.encoder[idx] = x
+
+        self.decoder = {}
+        for k, v in self.encoder.items():
+            self.decoder[v] = int(k)
+
+        self.trie = TRIE()
+        for t, i in self.decoder.items():
+            _ = self.trie.add(t, val=(t, i))
+        self.errors = errors  # how to handle errors in decoding
+        self.cache = {}
+        self.first_max_length = 0
+        super().__init__(
+            errors=errors,
+            **kwargs,
+        )
+
+    @property
+    def eos_token_id(self) -> Optional[int]:
+        return 0
+
+    @property
+    def eot_token_id(self) -> Optional[int]:
+        return 0
+
+    @property
+    def pad_token_id(self) -> Optional[int]:
+        return 0
+
+    @property
+    def vocab_size(self):
+        return len(self.encoder)
+
+    def get_vocab(self):
+        return dict(self.encoder, **self.added_tokens_encoder)
+
+    def add_tokens(self, new_tokens, special_tokens: bool = False):
+        for token in new_tokens:
+            token_id = self.convert_tokens_to_ids(token)
+            self.added_tokens_decoder[token_id] = token
```

Comment:
Is this to make it **faster** than using the original logic? The issue is that this does not seperate between the original tokens and the added tokens which are usually handled differently. This should be `add_tokens_to_vocab` to not change the API for our users ! 🤗

Reply:
Ok, that's right.